File: src/other/_Astart_2_4.c
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
const int mod=1e9+7;
const int N=1e5+5;
int a[N<<1];
ll num[N<<1];
ll dp[N<<1];

//*
ll ksm(ll a,ll b){
    ll s=1;
    while(b){
        if(b&1){
            a%=mod;
            s%=mod;
            s*=a;
        }
        a%=mod;
        a*=a;
        b>>=1;
    }
    return s%mod;
}
ll C(ll n,ll m){
    if(n==m||m==0) return 1;
    ll up=1,down=1;
    if(m>n-m) m=n-m;
    for(int i=0;i<m;i++){
        up=up*(n-i)%mod;
        down=down*(i+1)%mod;
    }
    return up*ksm(down,mod-2)%mod;
}
*/

// TLE
/*
const int MAXN = 1e5;
bool arr[MAXN+1] = {false};
vector<int> produce_prim_number(){
    vector<int> prim;
    prim.push_back(2);
    int i,j;
    for(i=3; i*i<=MAXN; i+=2){
        if(!arr[i]){
            prim.push_back(i);
            for(j=i*i; j<=MAXN; j+=i)
                arr[j] = true;
        }
    }
    while(i<=MAXN){
        if(!arr[i])
            prim.push_back(i);
        i+=2;
    }
    return prim;
}

//计算n!中素因子p的指数
int Cal(int x, int p){
    int ans = 0;
    long long rec = p;
    while(x>=rec){
        ans += x/rec;
        rec *= p;
    }
    return ans;
}

//计算n的k次方对M取模
int Pow(long long n, int k, int M){
    long long ans = 1;
    while(k){
        if(k&1){
            ans = (ans * n) % M;
        }
        n = (n * n) % M;
        k >>= 1;
    }
    return ans;
}

//计算C(n,m)
int C(int n, int m, vector<int> prim){
    long long ans = 1;
    int num;
    for(int i=0; i<prim.size() && prim[i]<=n; ++i){
        num = Cal(n, prim[i]) - Cal(m, prim[i]) - Cal(n-m, prim[i]);
        ans = (ans * Pow(prim[i], num, mod)) % mod;
    }
    return ans;
}
*/

int main(){
//    vector<int> prim = produce_prim_number();
    int T;
    scanf("%d",&T);
    while(T--){
        int n;
        scanf("%d",&n);
        int up=n<<1;
        for(int i=1;i<=up;i++){
            scanf("%d",&a[i]);
            num[i]=0;
        }
        for(int i=1;i<=up;i++) num[a[i]]++;
        dp[0]=1;
        ll sum=0;
        for(int i=1;i<=up;i++){
            if((sum&1)==0&&(num[i]&1)==0){ //偶偶
                dp[i]=dp[i-1]*C(num[i],num[i]>>1,prim)%mod;
            }
            else if((sum&1)==0&&(num[i]&1)==1){//偶奇
                dp[i]=dp[i-1]*C(num[i],num[i]>>1,prim)*2%mod;
            }
            else if((sum&1)==1&&(num[i]&1)==1){//奇奇
                dp[i]=dp[i-1]*C(num[i],num[i]>>1,prim)%mod;
            }
            else { //奇偶
                dp[i]=dp[i-1]*C(num[i],num[i]>>1,prim)%mod;
                if(num[i]){//不为0的偶数(即>=2)
                    dp[i]+=dp[i-1]*C(num[i],(num[i]-2)>>1,prim)%mod;
                    dp[i]%=mod;
                }
            }
            sum+=num[i];
        }
        printf("%lld\n",dp[up]);
    }
    return 0;
}

File: src/other/_Astart_2_4.java
package other;

import java.util.Scanner;

public class _Astart_2_4 {

    // TODO 未过
    // 估计原因 :: 求组合数出现问题
    // 求组合数的方法
    // 1) 利用递归式
    // 2) 利用乘法逆元
    // 3) 利用 分解成质因式
    // 4) 卢卡斯（用于求超大规模的素数求余数）

    private static final int VALUE = (int) (1e9 + 7);

    public static void main(String[] args) {
        make_p();
        Scanner scanner = new Scanner(System.in);
        int count = scanner.nextInt();
        int c = 0;
        int[] results = new int[count];
        while (c < count) {
            int len = scanner.nextInt();
            int[] as = new int[len * 2];
            for (int i = 0; i < len * 2; i++)
                as[i] = scanner.nextInt();
            long last = 1;
            int sum = 0;
            boolean lastStatus = true;
            for (int i = 0; i < as.length; i++) {
                int start = i;
                int currNum = as[i];
                int currCount;
                while (i < as.length && as[i] == currNum)
                    ++i;
                currCount = i - start;
                sum += currCount;
                boolean currStatus = (sum % 2) == 0;
                if (lastStatus && currStatus) {
                    last *= help(currCount, currCount / 2);
                } else if (!lastStatus && !currStatus) {
                    last *= (help(currCount, currCount / 2) + help(currCount, currCount / 2 - 1));
                } else {
                    last *= help(currCount, currCount / 2);
                    if (lastStatus)
                        last *= 2;
                }
                last %= VALUE;
                lastStatus = currStatus;
                --i;
            }
            results[c] = (int) last % VALUE;
            ++c;
        }
        for (int result : results)
            System.out.println(result);
    }

    private static final int MAX = 50000;
    private static int[] vis = new int[MAX + 1];
    private static int[] p = new int[MAX + 1];
    private static int tot = 0;

    // 挖素数
    private static void make_p() {
        vis[0] = vis[1] = 1;
        for (int i = 2; i <= MAX; i++)
            if (vis[i] == 0) {
                p[++tot] = i;
                for (int j = i * i; j > 0 && j <= MAX; j += i)
                    vis[j] = 1;
            }
    }

    // 快速幂
    private static long ksm(long a, long b) {
        long ans = 1, w = a;
        for (; b != 0; b >>= 1, w = (w * w) % VALUE)
            if ((b & 1) == 1)
                ans = (ans * w) % VALUE;
        return ans;
    }

    // 阶乘的标准分解式中素因数的指数
    private static long get(int x, int p) {
        long ans = 0;
        long rec = p;
        while (x >= rec) {
            ans += x / rec;
            rec *= p;
        }
        return ans;
    }

    private static int help(int x, int y) {
        if (x == 0 || y == 0)
            return 1;
        long ans = 1;
        for (int i = 1; i <= tot && p[i] <= x; i++) {
            long t = get(x, p[i]) - get(x - y, p[i]) - get(y, p[i]);
            ans = (ans * ksm(p[i], t)) % VALUE;
        }
        return (int) ans;
    }
}


// 方法一
// ERROR
    /*
    private static Map<Integer, Long> saveMap = new HashMap<>();

    private static long help(int big, int low) {
        if (low == 0)
            return 1;
        if (saveMap.get(low) == null) {
            long r = 1;
            for (int i = 2; i <= low; i++)
                r *= i;
            saveMap.put(low, r);
        }
        long d = saveMap.get(low);
        long m = 1;
        for (int i = 0; i < low; i++)
            m *= (big - i);
        return (m / d) % VALUE;
    }
    */

// 递归
// MLE
// 方法二
    /*
    private static Map<String, Integer> saveMap = new HashMap<>();

    private static int help(int big, int low) {
        if (big == 0 || low == 0)
            return 1;
        if (low > big)
            return 0;
        if (low == big)
            return 1;
        String key = big + "," + low;
        if (saveMap.get(key) != null)
            return saveMap.get(key);
        int result = (help(big - 1, low) + help(big - 1, low - 1)) % VALUE;
        saveMap.put(key, result);
        return result;
    }
    */

// 乘法逆元
// TLE
    /*
    private static long ksm(long a, long b) {
        long s = 1;
        while (b != 0) {
            if ((b & 1) != 0) {
                a %= VALUE;
                s %= VALUE;
                s *= a;
            }
            a %= VALUE;
            a *= a;
            b >>= 1;
        }
        return s % VALUE;
    }

    private static long help(long n, long m) {
        if (n == m || m == 0) return 1;
        long up = 1, down = 1;
        if (m > n - m) m = n - m;
        for (int i = 0; i < m; i++) {
            up = up * (n - i) % VALUE;
            down = down * (i + 1) % VALUE;
        }
        // (a/b)%mod=a*(b^(mod-2)) mod为素数
        return up * ksm(down, VALUE - 2) % VALUE;
    }
    */
